server/parsec/components/memory/datamodel.py

```diff
@@ -249,10 +253,149 @@ class MemoryInvitation:
     conduit_greeter_payload: bytes | None = field(default=None, repr=False)
     conduit_claimer_payload: bytes | None = field(default=None, repr=False)
 
+    # New fields for the new invitation system
+    # TODO: remove the old fields once the new system is fully deployed
+    greeting_sessions: dict[UserID, MemoryGreetingSession] = field(default_factory=dict)
+
     @property
     def is_deleted(self) -> bool:
         return self.deleted_on is not None
 
+    @property
+    def is_completed(self) -> bool:
+        return self.deleted_reason == MemoryInvitationDeletedReason.FINISHED
+
+    @property
+    def is_cancelled(self) -> bool:
+        return self.deleted_reason == MemoryInvitationDeletedReason.CANCELLED
+
+    def get_greeting_session(self, user_id: UserID) -> MemoryGreetingSession:
+        try:
+            return self.greeting_sessions[user_id]
+        except KeyError:
+            greeting_session = MemoryGreetingSession(
+                token=self.token,
+                greeter_id=user_id,
+            )
+            return self.greeting_sessions.setdefault(user_id, greeting_session)
+
+
+@dataclass(slots=True)
+class MemoryGreetingSession:
+    # Immutable properties
+    token: InvitationToken
+    greeter_id: UserID
+
+    # Mutable properties
+    greeting_attempts: list[GreetingAttemptID] = field(default_factory=list)
+
+    def get_active_greeting_attempt(self, org: MemoryOrganization) -> MemoryGreetingAttempt:
+        for attempt_id in self.greeting_attempts:
+            attempt = org.greeting_attempts[attempt_id]
+            if attempt.is_active():
+                return attempt
+        attempt = MemoryGreetingAttempt(
+            greeting_attempt=GreetingAttemptID.new(),
+            token=self.token,
+            greeter_id=self.greeter_id,
+        )
+        org.greeting_attempts[attempt.greeting_attempt] = attempt
+        self.greeting_attempts.append(attempt.greeting_attempt)
+        return attempt
+
+    def new_attempt_for_greeter(
+        self, org: MemoryOrganization, now: DateTime
+    ) -> MemoryGreetingAttempt:
+        while True:
```

Comment:
Should we add a limit for the number of retries, just in case ?

Reply:
In practice there could be at most one retry. I'll change the implementation to make it clearer.

Reply:
f79a54daf25763a1aa4140c5de40aa0190047647